File: 2022/Day02.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func score(in string, f func(a, b int) int) int {
	total := 0
	for _, l := range strings.Split(in, "\n") {
		if l == "" {
			continue
		}
		total += f(int(l[0]-'A'), int(l[2]-'X'))
	}
	return total
}

func part1(in string) int {
	return score(in, func(a, b int) int { return b + 1 + (b-a+4)%3*3 })
}

func part2(in string) int {
	return score(in, func(a, b int) int { return (a+b+2)%3 + 1 + b*3 })
}

func main() {
	b, _ := os.ReadFile("2022/Day02_input.txt")
	in := strings.TrimRight(string(b), "\n")
	fmt.Println("Part 1:", part1(in))
	fmt.Println("Part 2:", part2(in))
}

File: 2024/Day02.py
def safe(report):
    diffs = [b-a for a,b in zip(report, report[1:])]
    return all(1 <= d <= 3 for d in diffs) or all(-3 <= d <= -1 for d in diffs)

def part_1(reports):
    return sum(safe(r) for r in reports)

def part_2(reports):
    return sum(any(safe(r[:i] + r[i+1:]) for i in range(len(r))) for r in reports)

with open("2024/Day02_input.txt") as f:
    reports = [[int(v) for v in line.split()] for line in f.read().splitlines()]

    print(f"Part 1: {part_1(reports)}")
    print(f"Part 2: {part_2(reports)}")
